### PR TITLE
VITIS-7999 Update ReportDynamicRegions to use Hardware Contexts

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/VITIS-7999

Newer devices are able to support multiple xclbins, each taking up a single hardware context slot. The current report will only display the information for a single xclbin. Upgrade the report to allow the display multiple contexts while maintaining compatibility with older devices that do not have the concept of a hardware context.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Alveo devices only supported a single xclbin at any time. Newer devices like Versal and AIE support multiple xclbins. The current dynamic report will display only the first xclbin!

#### How problem was solved, alternative solutions (if any) and why they were rejected
Added a new query to gather all hardware contexts on a device. Refactored how the output json is generated to support multiple xclbins. Added support for legacy devices through the traditional queries.

#### Risks (if any) associated the changes in the commit
None. This change must be backward compatible.

#### What has been tested and how, request additional testing if necessary\
Ubuntu 20.04 U55C (Legacy path)
Command line output
```
dbenusov@xsjdbenusov50:/proj/rdi/staff/dbenusov$ xbutil examine -d 04:00 -r dynamic-regions -o json.json
WARNING: Unexpected xocl version (2.14.0) was found. Expected 2.15.0, to match XRT tools.

-------------------------------------------------
[0000:04:00.1] : xilinx_u55c_gen3x16_xdma_base_3
-------------------------------------------------

Xclbin UUID
  E106E953-CF90-4024-E075-282D1A7D820B

Compute Units
  PL Compute Units
    Index   Name                                              Base_Address    Usage   Status  
    0       verify:verify_1                                   0x800000        1       (IDLE)  

  PS Compute Units
    Index   Name                                              Base_Address    Usage   Status  

Successfully wrote the json file: json.json
```
json.json
```
dbenusov@xsjdbenusov50:/proj/rdi/staff/dbenusov$ cat json.json 
{
    "schema_version": {
        "schema": "JSON",
        "creation_date": "Thu Mar  9 18:49:07 2023 GMT"
    },
    "devices": [
        {
            "interface_type": "pcie",
            "device_id": "0000:04:00.1",
            "dynamic_regions": [
                {
                    "xclbin_uuid": "E106E953-CF90-4024-E075-282D1A7D820B",
                    "compute_units": [
                        {
                            "name": "verify:verify_1",
                            "base_address": "0x800000",
                            "usage": "1",
                            "type": "PL",
                            "status": {
                                "bit_mask": "0x4",
                                "bits_set": [
                                    "IDLE"
                                ]
                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```
Ubuntu 20.04 U55C (New path (Result was achieved by modifying the device query)
Command line output
```
dbenusov@xsjdbenusov50:/proj/rdi/staff/dbenusov$ xbutil examine -d 04:00 -r dynamic-regions -o json.json
WARNING: Unexpected xocl version (2.14.0) was found. Expected 2.15.0, to match XRT tools.

-------------------------------------------------
[0000:04:00.1] : xilinx_u55c_gen3x16_xdma_base_3
-------------------------------------------------

Xclbin UUID
  E106E953-CF90-4024-E075-282D1A7D820B

Compute Units
  PL Compute Units
    Index   Name                                              Base_Address    Usage   Status  
    0       verify:verify_1                                   0x800000        1       (IDLE)  

  PS Compute Units
    Index   Name                                              Base_Address    Usage   Status  

Xclbin UUID
  E106E953-CF90-4024-E075-282D1A7D820B

Compute Units
  PL Compute Units
    Index   Name                                              Base_Address    Usage   Status  
    0       verify:verify_1                                   0x800000        1       (IDLE)  

  PS Compute Units
    Index   Name                                              Base_Address    Usage   Status  

Successfully wrote the json file: json.json
```
json.json
```
dbenusov@xsjdbenusov50:/proj/rdi/staff/dbenusov$ cat json.json 
{
    "schema_version": {
        "schema": "JSON",
        "creation_date": "Thu Mar  9 19:12:58 2023 GMT"
    },
    "devices": [
        {
            "interface_type": "pcie",
            "device_id": "0000:04:00.1",
            "dynamic_regions": [
                {
                    "xclbin_uuid": "E106E953-CF90-4024-E075-282D1A7D820B",
                    "compute_units": [
                        {
                            "name": "verify:verify_1",
                            "base_address": "0x800000",
                            "usage": "1",
                            "type": "PL",
                            "status": {
                                "bit_mask": "0x4",
                                "bits_set": [
                                    "IDLE"
                                ]
                            }
                        }
                    ]
                },
                {
                    "xclbin_uuid": "E106E953-CF90-4024-E075-282D1A7D820B",
                    "compute_units": [
                        {
                            "name": "verify:verify_1",
                            "base_address": "0x800000",
                            "usage": "1",
                            "type": "PL",
                            "status": {
                                "bit_mask": "0x4",
                                "bits_set": [
                                    "IDLE"
                                ]
                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```
#### Documentation impact (if any)
None

### CODE REVIEW

File: src/runtime_src/core/common/info_memory.cpp [hunk 1]
```diff
@@ -506,34 +506,14 @@ get_ps_kernels(const xrt_core::device *device)
 }
 
 ptree_type
-populate_cus(const xrt_core::device *device)
+populate_cus(const xrt_core::device *device, std::vector<xq::kds_cu_info::data_type> cu_stats, std::vector<xq::kds_scu_info::data_type> scu_stats)
```
Comment (stsoe) [inline src/runtime_src/core/common/info_memory.cpp:509]: Oops by-value vectors?
```suggestion
populate_cus(const xrt_core::device *device, const std::vector<xq::kds_cu_info::data_type>& cu_stats, const std::vector<xq::kds_scu_info::data_type>& scu_stats)
```

File: src/runtime_src/core/common/info_memory.cpp [hunk 2]
```diff
@@ -587,18 +569,71 @@ populate_cus(const xrt_core::device *device)
...
   }
 
-  auto pt_dynamic_regions = xclbin_info(device);
-  pt_dynamic_regions.add_child("compute_units", pt);
-  return pt_dynamic_regions;
+  return pt;
+}
+
+static ptree_type
+populate_hardware_context(const xrt_core::device *device)
```
Comment (stsoe) [inline src/runtime_src/core/common/info_memory.cpp:576]:
```suggestion
populate_hardware_context(const xrt_core::device* device)
```

File: src/runtime_src/core/common/query_requests.h [hunk 3]
```diff
@@ -919,6 +920,23 @@ struct kds_scu_info : request
   get(const device*) const = 0;
 };
 
+
+struct hw_context_info : request
+{
+  struct data {
+    std::string xclbin_uuid;
+    kds_cu_info::result_type pl_compute_units;
+    kds_scu_info::result_type ps_compute_units;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:929]: Hmm, no suggestion here, but this seems a bit constrained.   Not all hw_context have pl cus or ps cus.

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/query_requests.h (hunk 3 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/query_requests.h:929]: The idea was if there are no ps or pl cus in a hardware context we just leave those fields empty. When a report iterates through these vectors they is parsed into the json, so if its empty no parsing! If AIE differs from PL compute units we can add a new vector for it or try to combine them all compute units into one. Might get a bit messy though.

Comment (stsoe) [reply src/runtime_src/core/common/query_requests.h:929]: Ok, so be it.  Please comment this struct and explain what it is for what the usage you just explained.

**Review by stsoe — APPROVED** [review]

File: src/runtime_src/core/common/info_memory.cpp [hunk 4]
```diff
@@ -506,34 +506,14 @@ get_ps_kernels(const xrt_core::device *device)
 }
 
 ptree_type
-populate_cus(const xrt_core::device *device)
+populate_cus(const xrt_core::device* device, const std::vector<xq::kds_cu_info::data_type>& cu_stats, const std::vector<xq::kds_scu_info::data_type>& scu_stats)
 {
   scheduler_update_stat(device);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/info_memory.cpp:511]: This call is no longer required as this is being called in populating hw context information.
All other changes are looking good.

Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/info_memory.cpp]: Good catch. Thank you.

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by chvamshi-xilinx** [comment]
retest this please

**Pull request merged**